earth_enterprise/src/common/SharedString.h

```diff
@@ -42,6 +43,7 @@ class SharedString {
         std::unordered_map<uint32_t, std::string> refFromKeyTable;
         std::unordered_map<std::string, uint32_t> keyFromRefTable;
         static uint32_t nextID;
+        mutable std::mutex string_storage_mutex_;
       public:
         const std::string & RefFromKey(const uint32_t &key) {
           auto refIter = refFromKeyTable.find(key);
```

Comment:
We need a lock everywhere that refFromKeyTable or keyFromRefTable is accessed in any way.. That includes this function and parts of the KeyFromRef function that are not currently locked.

Also, I agree with Stepehen that a read/write lock would be best here.